tests/compiler/test_plugins.py

```diff
@@ -435,6 +442,75 @@ def test_native_types(snippetcompiler: "SnippetCompilationTest") -> None:
 a = plugin_native_types::many_arguments(["a","c","b"], 1)
 
 none = plugin_native_types::as_none("a")
+
+# Union types (input)
+plugin_native_types::union_single_type(value="test")     # type value: Union[str]
+plugin_native_types::union_multiple_types(value="test")  # type value: Union[int, str]
+plugin_native_types::union_multiple_types(value=123)     # type value: Union[int, str]
+for val in ["test", 123, null]:
+    plugin_native_types::union_optional_1(value=val)     # type value: Union[None, int, str]
+    plugin_native_types::union_optional_2(value=val)     # type value: Optional[Union[int, str]]
+    plugin_native_types::union_optional_3(value=val)     # type value: Union[int, str] | None
+    plugin_native_types::union_optional_4(value=val)     # type value: None | Union[int, str]
+end
+
+# Union types (return value)
+plugin_native_types::union_return_single_type(value="test")     # type return value: Union[str]
+plugin_native_types::union_return_multiple_types(value="test")  # type return value: Union[str, int]
+plugin_native_types::union_return_multiple_types(value=123)     # type return value: Union[str, int]
+for val in ["test", 123, null]:
+    plugin_native_types::union_return_optional_1(value=val)     # type return value: Union[None, int, str]
+    plugin_native_types::union_return_optional_2(value=val)     # type return value: Optional[Union[int, str]]
+    plugin_native_types::union_return_optional_3(value=val)     # type return value: Union[int, str] | None
+    plugin_native_types::union_return_optional_4(value=val)     # type return value: None | Union[int, str]
+end
         """
     )
     compiler.do_compile()
+
+    # Parameter to plugin has incompatible type
+    for plugin_name, plugin_value, error_message in [
+        ("union_single_type", 123, "Invalid value '123', expected string"),
+        ("union_multiple_types", "[1, 2, 3]", "Invalid value '[1, 2, 3]', expected Union[int,string]"),
+        ("union_optional_1", 1.2, "Invalid value '1.2', expected Union[int,string]"),
+        ("union_optional_2", 1.2, "Invalid value '1.2', expected Union[int,string]"),
+        ("union_optional_3", 1.2, "Invalid value '1.2', expected Union[int,string]"),
+        ("union_optional_4", 1.2, "Invalid value '1.2', expected Union[int,string]"),
```

Comment:
These exception messages are in-line with what we already had, but I am wondering whether they are any good. It reports about an invalid subtype. For example it reports `Union[int, string]`, but shouldn't it report `Union[int, string]?` instead? Because the top-level type is optional. And wouldn't it be better to report the python type instead of the Inmanta type?

Reply:
I agree with both these statements

Reply:
> And wouldn't it be better to report the python type instead of the Inmanta type?

Tough one. On the one hand, it's more expressive and the plugin lives in the Python domain. On the other hand, the call originates from the DSL domain. I'm on the fence, leaning towards agreeing with you. @wouterdb wdyt?

Reply:
> but shouldn't it report Union[int, string]? instead?

I agree that would be better. Perhaps we can simply validate, and when it fails, construct a new exception with the Python type, instead of propagating the one raised by the DSL validation logic?